google/cloud/storage/internal/storage_metadata_decorator.h

```diff
@@ -0,0 +1,66 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_STORAGE_METADATA_DECORATOR_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_STORAGE_METADATA_DECORATOR_H
+
+#include "google/cloud/storage/internal/storage_stub.h"
+#include "google/cloud/version.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace storage {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+class StorageMetadata : public StorageStub {
+ public:
+  ~StorageMetadata() override = default;
+  explicit StorageMetadata(std::shared_ptr<StorageStub> child);
+
+  std::unique_ptr<google::cloud::internal::StreamingReadRpc<
+      google::storage::v2::ReadObjectResponse>>
+  ReadObject(std::unique_ptr<grpc::ClientContext> context,
+             google::storage::v2::ReadObjectRequest const& request) override;
+
+  std::unique_ptr<::google::cloud::internal::StreamingWriteRpc<
+      google::storage::v2::WriteObjectRequest,
+      google::storage::v2::WriteObjectResponse>>
+  WriteObject(std::unique_ptr<grpc::ClientContext> context) override;
+
+  StatusOr<google::storage::v2::StartResumableWriteResponse>
+  StartResumableWrite(
+      grpc::ClientContext& context,
+      google::storage::v2::StartResumableWriteRequest const& request) override;
+
+  StatusOr<google::storage::v2::QueryWriteStatusResponse> QueryWriteStatus(
+      grpc::ClientContext& context,
+      google::storage::v2::QueryWriteStatusRequest const& request) override;
+
+ private:
+  void SetMetadata(grpc::ClientContext& context,
+                   std::string const& request_params);
+  std::shared_ptr<StorageStub> child_;
+  std::string api_client_header_;
+};  // StorageMetadata
```

Comment:
(Not about this PR)

I am just now noticing that the generator adds this comment in the logging and metadata headers.... I don't think I like it....

https://github.com/googleapis/google-cloud-cpp/blob/354d2945c08b8308f57ea056502517d294e89d71/generator/internal/logging_decorator_generator.cc#L166
https://github.com/googleapis/google-cloud-cpp/blob/354d2945c08b8308f57ea056502517d294e89d71/generator/internal/metadata_decorator_generator.cc#L161

Reply:
Ack.  I am going to leave this as-is because I want the code to move towards what the generator creates.